✨ Support bytes in Options and Arguments

doubledare704 opened the pull request:
- Added support for bytes type in Typer by:
- Creating a custom BytesParamType class in typer/main.py
- Modifying the get_click_type function to handle bytes type
- Created comprehensive tests:
    - Basic tests for bytes type in tests/test_bytes_type.py
    - Advanced tests for base64 and hex encoding/decoding in tests/test_bytes_encoding.py
- Created example applications:
    - Simple bytes example in examples/bytes_type_example.py
    - Comprehensive encoding/decoding example in examples/bytes_encoding_example.py

Attempt to solve https://github.com/fastapi/typer/issues/536

doubledare704 committed:
Remove debug_process.py file

Add comprehensive tests for bytes type with base64 and hex encoding

Add support for bytes type in Options and Arguments

pre-commit-ci[bot] committed:
🎨 [pre-commit.ci] Auto format from pre-commit.com hooks

doubledare704 committed:
Fix linting issues and add type annotations

doubledare704 committed:
Fix mypy type error in BytesParamType.convert

pre-commit-ci[bot] committed:
🎨 [pre-commit.ci] Auto format from pre-commit.com hooks

doubledare704 committed:
make full coverage of tests

pre-commit-ci[bot] committed:
🎨 [pre-commit.ci] Auto format from pre-commit.com hooks

doubledare704 committed:
avoid assert false

doubledare704 committed:
avoid coverage in 1 line of tests

doubledare704 committed:
Merge branch 'master' into feature/support-bytes-type

doubledare704 committed:
Merge branch 'master' into feature/support-bytes-type

YuriiMotov reviewed (commented):
@doubledare704, thank you for your interest and efforts!

As I see it, the main reason to introduce `bytes` type support is to have the ability to use the same function outside the Typer, right?
Otherwise I don't see a point in this.. We can just accept `str` and decode it in one line. That would even be easier in cases where we want to let users specify the encoding as a CLI option

This implementation assumes that the argument value is encoded with system default encoding. It doesn't allow to specify the encoding.
This implementation assumes that the argument value is encoded with the system default encoding. It doesn’t allow specifying the encoding.
I believe we should at least allow specifying the encoding as a parameter to `Argument`/`Option`.

But first, I would wait for Sebastian to approve the idea

saky-semicolon reviewed (commented):
A few improvement opportunities:

Consider adding more edge case tests to further validate error handling.

In convert(), extracting decoding/encoding logic into helper functions could improve readability and reuse.

Documentation could be expanded with usage examples to make it easier for new users to try the features.

doubledare704 committed:
Merge branch 'fastapi:master' into feature/support-bytes-type

doubledare704 committed:
feat(bytes): support encoding/errors for bytes params\n\n- Add encoding-aware BytesParamType with helpers\n- Wire ParameterInfo.encoding/errors through Argument/Option\n- Add edge-case tests (non-ASCII, latin-1, ascii+errors, invalid encodings)\n- Expand examples to demonstrate encoding and errors\n- Add docs page: Tutorial -> CLI Parameter Types -> Bytes; wire into nav\n\nBackwards compatible: defaults to UTF-8 + 'strict' when not specified.

doubledare704 committed:
fix(tests): correct stderr assertion in bytes encoding test

- Update test to assert error message in `stderr` instead of `stdout`.
- Adjust indentation for function parameters to improve readability.

doubledare704 committed:
fix(tests): correct stderr assertion in bytes encoding test

- Update test to assert error message in `stderr` instead of `stdout`.
- Adjust indentation for function parameters to improve readability.

doubledare704 committed:
fix(tests): correct stderr assertion in bytes encoding test

- Update test to assert error message in `stderr` instead of `stdout`.
- Adjust indentation for function parameters to improve readability.

doubledare704 committed:
fix(tests): correct stderr assertion in bytes encoding test

- Update test to assert error message in `stderr` instead of `stdout`.
- Adjust indentation for function parameters to improve readability.

doubledare704 committed:
set no cover for uncoverable line

svlandeg committed:
update tutorial files to use Typer()

svlandeg committed:
Merge branch 'master' into feature/support-bytes-type

doubledare704 commented on docs/tutorial/parameter-types/bytes.md:
```diff
@@ -0,0 +1,38 @@
+# Bytes
```
seems like already has a link in mkdocs
mkdocs.yml:118

doubledare704 committed:
test(tutorial): add tests for bytes parameter types

- Add test cases for tutorials 001, 002, and 003 demonstrating bytes parameter usage.
- Verify CLI behavior and script-level execution with coverage checks.

doubledare704 committed:
docs(tutorial): fix range syntax for bytes tutorial example code

doubledare704 commented on tests/test_bytes_encoding.py:
```diff
@@ -0,0 +1,160 @@
+import base64
```
I've just added tests for tutorials. Please check.

svlandeg reviewed (commented):
@doubledare704: as Yurii said, we'll wait for further reviews until Tiangolo approves the idea. We have got this on our internal board, so no need for further pings 🙏

doubledare704 committed:
Merge branch 'master' into feature/support-bytes-type
